File: src/1-500/40/CombinationSumII.py
from typing import List


class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        res = []
        self.dfs(candidates, target, 0, res, [])
        return res

    def dfs(self, nums, target, index, res, path):
        if target < 0:
            return
        elif target == 0:
            res.append(path)
            return
        for i in range(index, len(nums)):
            if i > index and nums[i] == nums[i - 1]:
                continue
            self.dfs(nums, target - nums[i], i + 1, res, path + [nums[i]])




solution = Solution()
print(solution.combinationSum2([10,1,2,7,6,1,5], 8))

File: src/1-500/040/CombinationSumII.java
class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        List<List<Integer>> ans = new ArrayList<>();
        if (candidates == null || candidates.length == 0) {
            return ans;
        }
        Arrays.sort(candidates);
        buildList(ans, new ArrayList<>(), 0, target, candidates);
        return ans;
    }

    private void buildList(List<List<Integer>> ans, List<Integer> list, int index, int target, int[] candidates) {
        if (target < 0) {
            return;
        }
        if (target == 0) {
            ans.add(new ArrayList<>(list));
            return;
        }
        for (int i = index; i < candidates.length; i++) {
            if (i > index && candidates[i] == candidates[i- 1]) {
                continue;
            }
            if (target - candidates[i] < 0) {
                break;
            }
            list.add(candidates[i]);
            buildList(ans, list, i + 1, target - candidates[i], candidates);
            list.remove(list.size() - 1);
        }
        return;
    }
}